### ISSUE TITLE
Grid snapping

### ISSUE THREAD

**Issue by sutaburosu** [wokwi/wokwi-features#177]
Laying out diagram.json isn't easy, and expressing "busses" with wires routed neatly next to each other requires a calculator or fun mental arithmetic.

I feel there would be value in having grid-snapping.  So we might have a button toggleable between 3 or more states. I envisage 0.1" / 0.05" being the most useful,  which translates to multiples of 7.2 / 3.6 CSS units.

**Comment by urish** [comment]
Added initial support for grid-snapping - 0.1" grid. When routing wires, press the Shift key to snap.

**Comment by sutaburosu** [comment]
Thanks!  This is great so far.  The grid seems to be relative to the source pin.  Given the elements' pins may not align with a global grid, I think this is a good compromise.  It's certainly a lot less work to make a pretty diagram now, even if nudging slightly elements is necessary.

**Comment by urish** [comment]
One work around I was thinking about is snapping elements to grid based on their first pin when you shift-drag them. For most elements this would probably do the trick

**Comment by urish** [comment]
Experimental snapping implementation by @TimMathias: 

https://wokwi-diagram.vercel.app/arduino/new?template=arduino-uno

> There are two grid sizes: coarse (2.54 mm or 0.1") and fine (1.27 mm or 0.05"). Parts are snapped to the coarse grid only; new wires are snapped to the coarse grid by default, or the fine grid by pressing Shift, or off grid by pressing Alt.

Please share your feedback here

**Comment by TimMathias** [comment]
I've modified the snap-to-grid behaviour in response to comments and suggestions.

The grid snapping defaults to off to allow free movement. The Shift key temporarily switches to the coarse grid, and the Alt key and the Ctrl key temporarily switch to the fine grid.

The snap-to-grid mode is entered by pressing 'G' or by clicking the grid button. When in this mode, it defaults to the coarse grid. The Shift key temporarily switches to no grid (free movement), and the Alt key and the Ctrl key temporarily switch to the fine grid.

In other words, the Shift key temporarily toggles the grid snapping mode, and the Alt key and the Ctrl key always temporarily switch to the fine grid.

This grid snapping behaviour is the same for both parts and new wires.

Tick labels have been added to the rulers which default to units of millimetres, but you can switch to inches if you wish by clicking on the units in the top right corner.

**Comment by urish** [comment]
Closing - thanks @TimMathias for the hard work!

If there's still stuff missing, feel free to reopen / create a new issue
